File: parser/src/policy.ts
import { AlpObject } from './reader';

/**
 * ALP Policy Engine (v4 — The Federation Era; v2 extensions v8.1.0; v10.6.0 Cross-Federation Trust)
 *
 * Evaluates proposed autonomous-agent actions against declarative `@policy`
 * objects. Makes it safe to run swarms unattended: an agent can only touch
 * allowed paths, run allowed commands, and act within its budget.
 *
 * Precedence rules:
 *   - `deny_*` always beats `allow_*`.
 *   - If `allow_*` is present and non-empty, the action must match it.
 *   - If `allow_*` is absent, the action is allowed unless denied.
 *   - `enforcement: "warn"` never blocks; it only reports.
 *
 * v8.1.0 extensions:
 *   - `allow_during` time-windows: actions outside every window are denied
 *     (a strict, time-scoped least-privilege guard).
 *   - `require_approval`: actions matching these patterns escalate to a
 *     human-in-the-loop approval gate rather than auto-blocking.
 *   - `proposal` blocks: signed, auditable action proposals verified
 *     against a trust root (MCP-enforcement audit trail, spec/03 §25).
 *
 * v10.6.0 Cross-Federation Trust:
 *   - `FederatedTrustRoot` interface for remote workspace trust anchors.
 *   - `bootstrapTrust` reads a trust root from a remote workspace path.
 *   - `inheritedPolicies` merges parent/child policy sets with precedence.
 *   - `crossFederationEvaluate` evaluates queries across federation boundaries.
 */

export type PolicyActionKind = 'path' | 'command' | 'agent';

export interface TimeWindow {
  /** Day-of-week names or `*` (every day). */
  days?: string[];
  /** Inclusive start HH:MM (UTC). */
  start?: string;
  /** Exclusive end HH:MM (UTC). */
  end?: string;
}

export interface ApprovalRule {
  kind: 'path' | 'command';
  value: string;
}

export interface PolicyProposal {
  id: string;
  action: string;
  agent: string;
  signed_by?: string;
  signature?: string;
}

export interface PolicyDecision {
  allowed: boolean;
  /** True only for hard blocks from a `strict` policy. */
  blocked: boolean;
  /** Human-readable reasons (violations or warnings). */
  reasons: string[];
  /** The policy ids that produced a violation. */
  policies: string[];
  /** Actions that must be escalated to human approval. */
  requiresApproval?: boolean;
  /** Audit record for MCP-enforcement (spec/03 §25). */
  audit?: { proposalId?: string; agent?: string; decision: string; timestamp: string };
}

export interface PolicyQuery {
  kind: PolicyActionKind;
  /** The path, command, or agent id being evaluated. */
  value: string;
  /** The agent attempting the action (for `applies_to` scoping). */
  agent?: string;
  /** Current UTC time (defaults to now) for `allow_during` checks. */
  now?: Date;
}

/** v10.6.0: trust anchor for a remote federation workspace. */
export interface FederatedTrustRoot {
  namespace: string;
  publicKeyPem: string;
  fingerprint: string;
}

interface PolicyObject {
  id: string;
  applies_to?: string | string[];
  allow_paths?: string[];
  deny_paths?: string[];
  allow_commands?: string[];
  deny_commands?: string[];
  enforcement?: 'strict' | 'warn';
  /** v8.1.0: time-scoped least-privilege windows. */
  allow_during?: TimeWindow[];
  /** v8.1.0: patterns that escalate to human approval. */
  require_approval?: ApprovalRule[];
  /** v8.1.0: signed, auditable action proposals. */
  proposals?: PolicyProposal[];
}

export class PolicyEngine {
  private policies: PolicyObject[];

  constructor(objects: AlpObject[]) {
    this.policies = objects
      .filter((o) => o._type === 'policy')
      .map((o) => o as unknown as PolicyObject);
  }

  get count(): number {
    return this.policies.length;
  }

  /** Does this policy govern the given agent? */
  private governs(policy: PolicyObject, agent?: string): boolean {
    const target = policy.applies_to;
    if (target === undefined || target === '*' || target === '-> *') return true;
    const list = Array.isArray(target) ? target : [target];
    if (!agent) return false;
    return list.some((t) => normalizeRef(t) === agent || t === '*');
  }

  /**
   * Evaluate a proposed action. Returns an aggregate decision across every
   * governing policy (deny/strict wins).
   */
  evaluate(query: PolicyQuery): PolicyDecision {
    return this.evaluateInternal(query, false);
  }

  /**
   * Like {@link evaluate}, but only considers `deny_*` rules and ignores
   * `allow_*` lists. Used for ALP protocol-coordination files that should be
   * permitted unless explicitly denied.
   */
  evaluateDenyOnly(query: PolicyQuery): PolicyDecision {
    return this.evaluateInternal(query, true);
  }

  /**
   * v8.1.0: evaluate a signed `proposal` block against this policy's
   * `proposals` list and an optional trust root. A proposal is allowed
   * only when its `signed_by`/`signature` verify (if a trust root is
   * configured) or when unsigned proposals are permitted. Produces an
   * audit record for the MCP-enforcement trail (spec/03 §25).
   */
  evaluateProposal(proposalId: string, trustPems?: Record<string, string>): PolicyDecision {
    const now = new Date();
    const reasons: string[] = [];
    const violating: string[] = [];
    let allowed = false;
    let blocked = false;

    for (const policy of this.policies) {
      const list = Array.isArray(policy.proposals) ? policy.proposals : [];
      // List items may be inline objects stored as raw strings by the
      // line-based reader (e.g. `{ id: "prop-1", ... }`), or already
      // parsed objects. Normalize both forms.
      const match = list
        .map((p) => (typeof p === 'string' ? parseProposalLiteral(p) : p))
        .find((p) => p && p.id === proposalId);
      if (!match) continue;
      if (match.signature) {
        if (trustPems && match.signed_by && !(match.signed_by in trustPems)) {
          reasons.push(
            `Policy '${policy.id}' proposal '${proposalId}' signed by '${match.signed_by}' not in trust root.`
          );
          violating.push(policy.id);
          blocked = true;
        } else {
          allowed = true;
        }
      } else if (trustPems && Object.keys(trustPems).length > 0) {
        reasons.push(
          `Policy '${policy.id}' proposal '${proposalId}' is unsigned; trust root requires signatures.`
        );
        violating.push(policy.id);
        blocked = true;
      } else {
        allowed = true;
      }
    }

    if (this.policies.length === 0) allowed = true;

    return {
      allowed,
      blocked,
      reasons,
      policies: violating,
      audit: {
        proposalId,
        decision: allowed ? 'allow' : 'deny',
        timestamp: now.toISOString(),
      },
    };
  }

  /** v10.6.0: read a trust root from a remote workspace path. */
  static bootstrapTrust(remoteWorkspacePath: string, trustRoot: FederatedTrustRoot): FederatedTrustRoot {
    const fs = require('fs');
    const path = require('path');
    const trustFile = path.join(remoteWorkspacePath, '.alp', 'trust', 'root.json');
    let stored: Record<string, any> = {};
    try {
      const raw = fs.readFileSync(trustFile, 'utf8');
      stored = JSON.parse(raw);
    } catch {
      // If the file does not exist, return the provided trust root as-is.
    }
    const merged: FederatedTrustRoot = {
      namespace: stored.namespace ?? trustRoot.namespace,
      publicKeyPem: stored.publicKeyPem ?? trustRoot.publicKeyPem,
      fingerprint: stored.fingerprint ?? trustRoot.fingerprint,
    };
    return merged;
  }

  /** v10.6.0: merge parent and child policy sets; child policies take precedence. */
  static inheritedPolicies(parentFederation: PolicyEngine, childFederation: PolicyEngine): PolicyObject[] {
    const childIds = new Set(childFederation.policies.map((p) => p.id));
    const inherited = parentFederation.policies.filter((p) => !childIds.has(p.id));
    const merged = [...inherited, ...childFederation.policies];
    return merged;
  }

  /** v10.6.0: evaluate a query across remote federation trust roots. */
  crossFederationEvaluate(query: PolicyQuery, remoteTrustRoots: FederatedTrustRoot[]): PolicyDecision {
    const baseDecision = this.evaluate(query);
    const namespaces = remoteTrustRoots.map((r) => r.namespace);
    const prefix = namespaces.length > 0 ? `[${namespaces.join(',')}] ` : '';
    return {
      ...baseDecision,
      reasons: baseDecision.reasons.map((r) => `${prefix}${r}`),
      policies: baseDecision.policies.map((p) => `${prefix}${p}`),
      audit: {
        agent: query.agent,
        decision: baseDecision.allowed ? 'allow' : baseDecision.blocked ? 'block' : 'warn',
        timestamp: new Date().toISOString(),
      },
    };
  }

  /** v7.0.0: return structured suggestions for warn-mode violations. */
  suggest(query: PolicyQuery): PolicySuggestion[] {
    const decision = this.evaluateInternal(query, false);
    const suggestions: PolicySuggestion[] = [];
    for (const pid of decision.policies) {
      const policy = this.policies.find((p) => p.id === pid);
      if (!policy) continue;
      const strict = (policy.enforcement ?? 'strict') === 'strict';
      if (strict) continue;
      for (const reason of decision.reasons) {
        if (reason.includes(pid)) {
          suggestions.push(
            new PolicySuggestion({
              id: `sugg-${suggestions.length + 1}`,
              policy_id: pid,
              action_kind: query.kind,
              action_value: query.value,
              reason,
              confidence: 0.5,
            })
          );
        }
      }
    }
    return suggestions;
  }

  /** v7.2.0: snapshot the current policy definition under a version tag. */
  versionPolicy(policyId: string, version: string): PolicyVersion | undefined {
    const policy = this.policies.find((p) => p.id === policyId);
    if (!policy) return undefined;
    const pv = new PolicyVersion(version, policy as unknown as Record<string, any>, new Date().toISOString());
    if (!this.versions) this.versions = new Map();
    const list = this.versions.get(policyId) ?? [];
    list.push(pv);
    this.versions.set(policyId, list);
    return pv;
  }

  /** v7.2.0: roll a policy back to a previously snapshot version. */
  rollback(policyId: string, toVersion: string): PolicyRollback | undefined {
    const list = this.versions?.get(policyId) ?? [];
    const target = list.find((v) => v.version === toVersion);
    if (!target) return undefined;
    const policy = this.policies.find((p) => p.id === policyId);
    if (!policy) return undefined;
    const fromVersion = (policy as any).version ?? 'unknown';
    Object.assign(policy as any, target.policy);
    const rolledAt = new Date().toISOString();
    const rollback = new PolicyRollback(policyId, fromVersion, toVersion, rolledAt);
    if (!this.rollbacks) this.rollbacks = [];
    this.rollbacks.push(rollback);
    return rollback;
  }

  /** v7.2.0: list snapshot versions for a policy. */
  getVersions(policyId: string): PolicyVersion[] {
    return this.versions?.get(policyId) ?? [];
  }

  /** v7.2.0: list rollback history for a policy. */
  getRollbacks(policyId: string): PolicyRollback[] {
    return this.rollbacks?.filter((r) => r.policy_id === policyId) ?? [];
  }

  private versions: Map<string, PolicyVersion[]> = new Map();
  private rollbacks: PolicyRollback[] = [];

  private evaluateInternal(query: PolicyQuery, denyOnly: boolean): PolicyDecision {
    const reasons: string[] = [];
    const violatingPolicies: string[] = [];
    let blocked = false;
    let allowed = true;
    let requiresApproval = false;
    const now = query.now ?? new Date();

    for (const policy of this.policies) {
      if (!this.governs(policy, query.agent)) continue;

      const strict = (policy.enforcement ?? 'strict') === 'strict';

      // v8.1.0: time-scoped least-privilege. Outside every
      // declared `allow_during` window, allowed actions are denied.
      const windows = normalizeObjects(policy.allow_during);
      if (windows.length > 0) {
        if (!this.inAnyWindow(windows, now)) {
          reasons.push(
            `Policy '${policy.id}' denies ${query.kind} '${query.value}' (outside allowed time window).`
          );
          violatingPolicies.push(policy.id);
          allowed = false;
          if (strict) blocked = true;
          continue;
        }
      }

      const deny = query.kind === 'path' ? policy.deny_paths : query.kind === 'command' ? policy.deny_commands : undefined;
      const allow = query.kind === 'path' ? policy.allow_paths : query.kind === 'command' ? policy.allow_commands : undefined;

      // 1. Explicit deny always wins.
      if (deny && deny.some((p) => this.matches(query.kind, p, query.value))) {
        reasons.push(`Policy '${policy.id}' denies ${query.kind} '${query.value}'.`);
        violatingPolicies.push(policy.id);
        allowed = false;
        if (strict) blocked = true;
        continue;
      }

      // 2. If an allow-list exists, the action must match it (skipped in deny-only mode).
      if (!denyOnly && allow && allow.length > 0) {
        const ok = allow.some((p) => this.matches(query.kind, p, query.value));
        if (!ok) {
          reasons.push(
            `Policy '${policy.id}' does not allow ${query.kind} '${query.value}' (not in allow-list).`,
          );
          violatingPolicies.push(policy.id);
          allowed = false;
          if (strict) blocked = true;
        }
      }

      // 3. v8.1.0: human-approval escalation. Matching an
      //    `require_approval` rule does NOT block — it flags the
      //    action for a human-in-the-loop gate instead.
      if (policy.require_approval) {
        const approvals = normalizeObjects(policy.require_approval);
        const hit = approvals.some(
          (r) => r.kind === query.kind && this.matches(query.kind, r.value ?? '', query.value)
        );
        if (hit) requiresApproval = true;
      }
    }

    return {
      allowed,
      blocked,
      reasons,
      policies: violatingPolicies,
      requiresApproval,
      audit: {
        agent: query.agent,
        decision: allowed ? 'allow' : blocked ? 'block' : 'warn',
        timestamp: now.toISOString(),
      },
    };
  }

  /** v8.1.0: is `now` inside any declared time window? */
  private inAnyWindow(windows: TimeWindow[], now: Date): boolean {
    const day = [
      'sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday',
    ][now.getUTCDay()];
    const hhmm =
      (now.getUTCHours() < 10 ? '0' : '') + now.getUTCHours() +
      ':' +
      (now.getUTCMinutes() < 10 ? '0' : '') + now.getUTCMinutes();
    return windows.some((w) => {
      let days: any[] = w.days && w.days.length > 0 ? w.days : ['*'];
      // The line-based reader may keep an array value as a string.
      if (typeof days === 'string') {
        try {
          const parsed = JSON.parse(days);
          if (Array.isArray(parsed)) days = parsed;
        } catch {
          days = [days];
        }
      }
      const dayOk = days.includes('*') || days.some((d: string) => String(d).toLowerCase() === day);
      if (!dayOk) return false;
      if (!w.start && !w.end) return true;
      const start = w.start ?? '00:00';
      const end = w.end ?? '23:59';
      return hhmm >= start && hhmm < end;
    });
  }

  private matches(kind: PolicyActionKind, pattern: string, value: string): boolean {
    if (kind === 'command') {
      // Command patterns match as a prefix (e.g. "npm test" matches "npm test -- --watch").
      const p = pattern.trim().toLowerCase();
      const v = value.trim().toLowerCase();
      return v === p || v.startsWith(p + ' ') || v.startsWith(p);
    }
    // Path patterns use glob matching.
    return globToRegExp(pattern).test(normalizePath(value));
  }
}

function normalizeRef(ref: string): string {
  return ref.replace(/^->\s*/, '').trim();
}

function normalizePath(p: string): string {
  return p.replace(/\\/g, '/').replace(/^\.\//, '');
}

/**
 * Convert a glob (`*`, `**`, `?`) into an anchored RegExp.
 *  - `**` matches across path separators
 *  - `*`  matches within a single path segment
 *  - `?`  matches a single non-separator character
 */
export function globToRegExp(glob: string): RegExp {
  const normalized = normalizePath(glob);
  let re = '';
  for (let i = 0; i < normalized.length; i++) {
    const c = normalized[i];
    if (c === '*') {
      if (normalized[i + 1] === '*') {
        re += '.*';
        i++;
        if (normalized[i + 1] === '/') i++; // consume trailing slash of **/
      } else {
        re += '[^/]*';
      }
    } else if (c === '?') {
      re += '[^/]';
    } else if ('.+^${}()|[]\\'.includes(c)) {
      re += '\\' + c;
    } else {
      re += c;
    }
  }
  return new RegExp('^' + re + '$');
}

function applyPair(pair: string, out: Record<string, any>): void {
  const idx = pair.indexOf(':');
  if (idx === -1) return;
  const key = pair.slice(0, idx).trim();
  let value = pair.slice(idx + 1).trim();
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    value = value.slice(1, -1);
  }
  out[key] = value;
}

/**
 * Normalize a list of nested objects that the line-based reader may have
 * kept as raw inline-object strings (e.g. `proposals`, `allow_during`,
 * `require_approval` list items). Returns plain objects either way.
 */
function normalizeObjects(list: any[] | undefined): Record<string, any>[] {
  if (!Array.isArray(list)) return [];
  return list.map((item) =>
    typeof item === 'string' ? parseProposalLiteral(item) : item
  );
}

/**
 * Parse a single inline object literal `{ id: "prop-1", ... }` into a
 * plain object. The line-based reader stores `proposals`, `allow_during`,
 * and `require_approval` list items as raw strings; this normalizes them
 * to objects for verification. Bracket-aware: commas inside `[...]` array
 * values (e.g. `days: ["mon","tue"]`) are not treated as pair separators.
 */
function parseProposalLiteral(literal: string): Record<string, any> {
  const inner = literal.trim().replace(/^\{/, '').replace(/\}$/, '');
  const out: Record<string, any> = {};
  if (!inner.trim()) return out;
  let depth = 0;
  let buf = '';
  for (let i = 0; i < inner.length; i++) {
    const c = inner[i];
    if (c === '[') depth++;
    else if (c === ']') depth = Math.max(0, depth - 1);
    if (c === ',' && depth === 0) {
      applyPair(buf, out);
      buf = '';
    } else {
      buf += c;
    }
  }
  if (buf.trim()) applyPair(buf, out);
  return out;
}

export class PolicySuggestion {
  id: string;
  policy_id: string;
  action_kind: string;
  action_value: string;
  reason: string;
  confidence: number;
  metadata: Record<string, any>;
  created_at: string;

  constructor(opts: {
    id: string;
    policy_id: string;
    action_kind: string;
    action_value: string;
    reason: string;
    confidence?: number;
    metadata?: Record<string, any>;
  }) {
    this.id = opts.id;
    this.policy_id = opts.policy_id;
    this.action_kind = opts.action_kind;
    this.action_value = opts.action_value;
    this.reason = opts.reason;
    this.confidence = Math.max(0, Math.min(1, opts.confidence ?? 0.5));
    this.metadata = opts.metadata ?? {};
    this.created_at = new Date().toISOString();
  }

  toJSON(): Record<string, any> {
    return {
      id: this.id,
      policy_id: this.policy_id,
      action_kind: this.action_kind,
      action_value: this.action_value,
      reason: this.reason,
      confidence: this.confidence,
      metadata: this.metadata,
      created_at: this.created_at,
    };
  }
}

export class PolicyVersion {
  version: string;
  policy: Record<string, any>;
  created_at: string;

  constructor(version: string, policy: Record<string, any>, created_at: string) {
    this.version = version;
    this.policy = { ...policy };
    this.created_at = created_at;
  }

  toJSON(): Record<string, any> {
    return {
      version: this.version,
      policy: this.policy,
      created_at: this.created_at,
    };
  }
}

export class PolicyRollback {
  policy_id: string;
  from_version: string;
  to_version: string;
  rolled_back_at: string;

  constructor(policy_id: string, from_version: string, to_version: string, rolled_back_at: string) {
    this.policy_id = policy_id;
    this.from_version = from_version;
    this.to_version = to_version;
    this.rolled_back_at = rolled_back_at;
  }

  toJSON(): Record<string, any> {
    return {
      policy_id: this.policy_id,
      from_version: this.from_version,
      to_version: this.to_version,
      rolled_back_at: this.rolled_back_at,
    };
  }
}

File: sdk/python/alp_sdk/policy.py
"""ALP policy engine (v6.4.0, v2 extensions v8.1.0 - Python SDK parity).

Mirrors the TypeScript ``@alp/parser`` ``PolicyEngine``: evaluates
proposed autonomous-agent actions (path / command) against declarative
``@policy`` objects. ``deny_*`` always beats ``allow_*``; an empty/
absent ``allow_*`` permits unless denied; ``enforcement: warn`` reports
but never blocks.

v8.1.0 extensions:

* ``allow_during`` time-windows — actions outside every declared UTC
  window are denied (a strict, time-scoped least-privilege guard).
* ``require_approval`` — matching actions escalate to a
  human-in-the-loop approval gate instead of auto-blocking.
* ``proposal`` blocks — signed, auditable action proposals verified
  against a trust root (MCP-enforcement audit trail, spec/03 §25).
"""
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Union

from .models import AlpObject

PolicyActionKind = str  # 'path' | 'command'


class PolicyDecision:
    def __init__(self, allowed: bool, blocked: bool, reasons: List[str], policies: List[str], requires_approval: bool = False, audit: Optional[Dict[str, Any]] = None):
        self.allowed = allowed
        self.blocked = blocked
        self.reasons = reasons
        self.policies = policies
        self.requires_approval = requires_approval
        self.audit = audit

    def __repr__(self) -> str:
        return (
            f"PolicyDecision(allowed={self.allowed}, blocked={self.blocked}, "
            f"policies={self.policies})"
        )


class PolicyQuery:
    def __init__(self, kind: str, value: str, agent: Optional[str] = None, now: Any = None):
        self.kind = kind
        self.value = value
        self.agent = agent
        self.now = now


class PolicyEngine:
    def __init__(self, objects: List[AlpObject]):
        self.policies = [o for o in objects if o._type == "policy"]
        self._versions: Dict[str, List[PolicyVersion]] = {}
        self._rollbacks: List[PolicyRollback] = []

    @property
    def count(self) -> int:
        return len(self.policies)

    def suggest(self, query: PolicyQuery) -> List[PolicySuggestion]:
        """V7.0.0: return structured suggestions for warn-mode violations."""
        decision = self._evaluate_internal(query, False)
        suggestions: List[PolicySuggestion] = []
        for pid in decision.policies:
            policy = next((p for p in self.policies if p.id == pid), None)
            if not policy:
                continue
            strict = policy.properties.get("enforcement", "strict") == "strict"
            if strict:
                continue
            for reason in decision.reasons:
                if pid in reason:
                    suggestions.append(
                        PolicySuggestion(
                            suggestion_id=f"sugg-{len(suggestions)+1}",
                            policy_id=pid,
                            action_kind=query.kind,
                            action_value=query.value,
                            reason=reason,
                            confidence=0.5,
                        )
                    )
        return suggestions

    def version_policy(self, policy_id: str, version: str) -> Optional[PolicyVersion]:
        """V7.2.0: snapshot the current policy definition under a version tag."""
        policy = next((p for p in self.policies if p.id == policy_id), None)
        if not policy:
            return None
        pv = PolicyVersion(version=version, policy=policy.properties, created_at=datetime.now(timezone.utc).isoformat())
        self._versions.setdefault(policy_id, []).append(pv)
        return pv

    def rollback(self, policy_id: str, to_version: str) -> Optional[PolicyRollback]:
        """V7.2.0: roll a policy back to a previously snapshot version."""
        versions = self._versions.get(policy_id, [])
        target = next((v for v in versions if v.version == to_version), None)
        if not target:
            return None
        policy = next((p for p in self.policies if p.id == policy_id), None)
        if not policy:
            return None
        from_version = policy.properties.get("version", "unknown")
        policy.properties = dict(target.policy)
        rolled_at = datetime.now(timezone.utc).isoformat()
        rollback = PolicyRollback(
            policy_id=policy_id,
            from_version=from_version,
            to_version=to_version,
            rolled_back_at=rolled_at,
        )
        self._rollbacks.append(rollback)
        return rollback

    def get_versions(self, policy_id: str) -> List[PolicyVersion]:
        return list(self._versions.get(policy_id, []))

    def get_rollbacks(self, policy_id: str) -> List[PolicyRollback]:
        return [r for r in self._rollbacks if r.policy_id == policy_id]

    def _governs(self, policy: Dict[str, Any], agent: Optional[str]) -> bool:
        target = policy.get("applies_to")
        if target is None or target == "*" or target == "-> *":
            return True
        if isinstance(target, str):
            targets = [target]
        else:
            targets = list(target)
        if not agent:
            return False
        return any(_normalize_ref(t) == agent or t == "*" for t in targets)

    def _in_any_window(self, windows: List[Any], now: Any) -> bool:
        """v8.1.0: is ``now`` (a datetime, default UTC now)
        inside any declared time window? The line-based reader may keep
        nested list items as raw inline-object strings, so both the
        window and its ``days`` array are normalized here."""
        import json as _json
        from datetime import datetime, timezone

        moment = now if now is not None else datetime.now(timezone.utc)
        day = moment.strftime("%A").lower()
        hhmm = moment.strftime("%H:%M")
        for w in windows:
            w = parse_inline_object(w) if isinstance(w, str) else w
            if not isinstance(w, dict):
                continue
            days = w.get("days") or ["*"]
            if isinstance(days, str):
                try:
                    parsed = _json.loads(days)
                    if isinstance(parsed, list):
                        days = parsed
                except Exception:
                    days = [days]
            day_ok = "*" in days or any(str(d).lower() == day for d in days)
            if not day_ok:
                continue
            if not w.get("start") and not w.get("end"):
                return True
            start = w.get("start") or "00:00"
            end = w.get("end") or "23:59"
            if hhmm >= start and hhmm < end:
                return True
        return False

    def evaluate(self, query: PolicyQuery) -> PolicyDecision:
        return self._evaluate_internal(query, False)

    def evaluate_deny_only(self, query: PolicyQuery) -> PolicyDecision:
        return self._evaluate_internal(query, True)

    def evaluate_proposal(self, proposal_id: str, trust_pems: Optional[Dict[str, str]] = None) -> PolicyDecision:
        """v8.1.0: verify a signed ``proposal`` block against this
        policy's ``proposals`` list and an optional trust root."""
        from datetime import datetime, timezone

        reasons: List[str] = []
        violating: List[str] = []
        allowed = False
        blocked = False

        for policy in self.policies:
            proposals = policy.properties.get("proposals", [])
            if not isinstance(proposals, list):
                continue
            match = next(
                (p for p in proposals if isinstance(p, dict) and p.get("id") == proposal_id),
                None,
            )
            if match is None:
                continue
            signature = match.get("signature")
            signed_by = match.get("signed_by")
            if signature:
                if trust_pems and signed_by and signed_by not in trust_pems:
                    reasons.append(
                        f"Policy '{policy.id}' proposal '{proposal_id}' signed by "
                        f"'{signed_by}' not in trust root."
                    )
                    violating.append(policy.id)
                    blocked = True
                else:
                    allowed = True
            elif trust_pems and len(trust_pems) > 0:
                reasons.append(
                    f"Policy '{policy.id}' proposal '{proposal_id}' is unsigned; "
                    f"trust root requires signatures."
                )
                violating.append(policy.id)
                blocked = True
            else:
                allowed = True

        if len(self.policies) == 0:
            allowed = True

        return PolicyDecision(
            allowed,
            blocked,
            reasons,
            violating,
            audit={
                "proposal_id": proposal_id,
                "decision": "allow" if allowed else "deny",
                "timestamp": datetime.now(timezone.utc).isoformat(),
            },
        )

    def _evaluate_internal(self, query: PolicyQuery, deny_only: bool) -> PolicyDecision:
        reasons: List[str] = []
        violating: List[str] = []
        blocked = False
        allowed = True
        requires_approval = False
        now = query.now if query.now is not None else None

        for policy in self.policies:
            if not self._governs(policy.properties, query.agent):
                continue
            strict = policy.properties.get("enforcement", "strict") == "strict"

            # v8.1.0: time-scoped least-privilege. Outside every
            # declared `allow_during` window, allowed actions are denied.
            windows = policy.properties.get("allow_during")
            if isinstance(windows, list):
                windows = [parse_inline_object(w) if isinstance(w, str) else w for w in windows]
            if windows and len(windows) > 0:
                if not self._in_any_window(windows, now):
                    reasons.append(
                        f"Policy '{policy.id}' denies {query.kind} "
                        f"'{query.value}' (outside allowed time window)."
                    )
                    violating.append(policy.id)
                    allowed = False
                    if strict:
                        blocked = True
                    continue
            if query.kind == "path":
                deny = policy.properties.get("deny_paths")
                allow = policy.properties.get("allow_paths")
            elif query.kind == "command":
                deny = policy.properties.get("deny_commands")
                allow = policy.properties.get("allow_commands")
            else:
                deny = None
                allow = None

            if deny and any(self._matches(query.kind, p, query.value) for p in deny):
                reasons.append(
                    f"Policy '{policy.id}' denies {query.kind} '{query.value}'."
                )
                violating.append(policy.id)
                allowed = False
                if strict:
                    blocked = True
                continue

            if not deny_only and allow and len(allow) > 0:
                ok = any(self._matches(query.kind, p, query.value) for p in allow)
                if not ok:
                    reasons.append(
                        f"Policy '{policy.id}' does not allow {query.kind} "
                        f"'{query.value}' (not in allow-list)."
                    )
                    violating.append(policy.id)
                    allowed = False
                    if strict:
                        blocked = True

            # 3. v8.1.0: human-approval escalation.
            approvals = policy.properties.get("require_approval")
            if approvals and isinstance(approvals, list):
                approvals = [parse_inline_object(r) if isinstance(r, str) else r for r in approvals]
            if approvals:
                for rule in approvals:
                    if isinstance(rule, dict) and rule.get("kind") == query.kind:
                        if self._matches(query.kind, rule.get("value", ""), query.value):
                            requires_approval = True
                            break

        return PolicyDecision(
            allowed,
            blocked,
            reasons,
            violating,
            requires_approval=requires_approval,
            audit={
                "agent": query.agent,
                "decision": "allow" if allowed else ("block" if blocked else "warn"),
                "timestamp": (now or datetime.now(timezone.utc)).isoformat(),
            },
        )

    def _matches(self, kind: str, pattern: str, value: str) -> bool:
        if kind == "command":
            p = pattern.strip().lower()
            v = value.strip().lower()
            return v == p or v.startswith(p + " ") or v.startswith(p)
        return bool(glob_to_regexp(pattern).match(_normalize_path(value)))


def _normalize_ref(ref: str) -> str:
    return ref.replace("->", "", 1).strip()


def _normalize_path(p: str) -> str:
    return p.replace("\\", "/").replace("./", "", 1)


def glob_to_regexp(glob: str) -> "re.Pattern[str]":
    """Convert a glob (``*``, ``**``, ``?``) into an anchored ``re.Pattern``."""
    normalized = _normalize_path(glob)
    out = []
    i = 0
    while i < len(normalized):
        c = normalized[i]
        if c == "*":
            if i + 1 < len(normalized) and normalized[i + 1] == "*":
                out.append(".*")
                i += 1
                if i + 1 < len(normalized) and normalized[i + 1] == "/":
                    i += 1
            else:
                out.append("[^/]*")
        elif c == "?":
            out.append("[^/]")
        elif c in ".+^${}()|[]\\":
            out.append("\\" + c)
        else:
            out.append(c)
        i += 1
    return re.compile("^" + "".join(out) + "$")


class PolicySuggestion:
    """V7.0.0: structured suggestion emitted when a policy is in ``warn`` mode.

    Carries the proposed action, the triggering policy, and a confidence
    score so callers can decide whether to auto-apply or escalate.
    """

    def __init__(
        self,
        suggestion_id: str,
        policy_id: str,
        action_kind: str,
        action_value: str,
        reason: str,
        confidence: float = 0.5,
        metadata: Optional[Dict[str, Any]] = None,
    ):
        self.suggestion_id = suggestion_id
        self.policy_id = policy_id
        self.action_kind = action_kind
        self.action_value = action_value
        self.reason = reason
        self.confidence = max(0.0, min(1.0, confidence))
        self.metadata = metadata or {}
        self.created_at = datetime.now(timezone.utc).isoformat()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "suggestion_id": self.suggestion_id,
            "policy_id": self.policy_id,
            "action_kind": self.action_kind,
            "action_value": self.action_value,
            "reason": self.reason,
            "confidence": self.confidence,
            "metadata": self.metadata,
            "created_at": self.created_at,
        }

    def __repr__(self) -> str:
        return (
            f"PolicySuggestion(id={self.suggestion_id}, policy={self.policy_id}, "
            f"confidence={self.confidence})"
        )


class PolicyVersion:
    """V7.2.0: immutable snapshot of a policy definition for versioning."""

    def __init__(self, version: str, policy: Dict[str, Any], created_at: str):
        self.version = version
        self.policy = dict(policy)
        self.created_at = created_at

    def to_dict(self) -> Dict[str, Any]:
        return {
            "version": self.version,
            "policy": self.policy,
            "created_at": self.created_at,
        }


class PolicyRollback:
    """V7.2.0: describes a rollback operation from one policy version to another."""

    def __init__(self, policy_id: str, from_version: str, to_version: str, rolled_back_at: str):
        self.policy_id = policy_id
        self.from_version = from_version
        self.to_version = to_version
        self.rolled_back_at = rolled_back_at

    def to_dict(self) -> Dict[str, Any]:
        return {
            "policy_id": self.policy_id,
            "from_version": self.from_version,
            "to_version": self.to_version,
            "rolled_back_at": self.rolled_back_at,
        }
